File: source/ocr/src/merchant_normalizer.py
"""
Merchant name normalization for the expense-tracker OCR worker.

Algorithm must produce IDENTICAL output to the .NET MerchantNormalizer.cs
in ExpenseTracker.Shared — verified by shared test fixtures at
tests/merchant_normalization_fixtures.json.

Rules:
  1. Lowercase
  2. Remove punctuation characters: . , ' " - _ & /
  3. Collapse whitespace to single space
  4. Strip leading/trailing whitespace
"""

import re

_PUNCTUATION_RE = re.compile(r"[.,'\"\-_&/]")
_WHITESPACE_RE = re.compile(r"\s+")


def normalize_merchant(name: str | None) -> str:
    """Return the canonical normalized form of a merchant name.

    Returns an empty string for None or whitespace-only input.
    """
    if not name or not name.strip():
        return ""

    result = name.lower()
    result = _PUNCTUATION_RE.sub(" ", result)
    result = _WHITESPACE_RE.sub(" ", result)
    return result.strip()

File: source/api/src/ExpenseTracker.Shared/MerchantNormalizer.cs
using System.Text.RegularExpressions;

namespace ExpenseTracker.Shared;

/// <summary>
/// Normalizes merchant names to a canonical lowercase form used for
/// category mapping, duplicate detection, and tag history lookups.
/// The algorithm must produce identical output to the Python OCR worker's
/// normalize_merchant() function — verified by shared test fixtures.
/// </summary>
public static partial class MerchantNormalizer
{
    // Remove punctuation characters: . , ' " - _ & /
    [GeneratedRegex(@"[.,'""\-_&/]", RegexOptions.Compiled)]
    private static partial Regex PunctuationPattern();

    // Collapse multiple whitespace characters into a single space
    [GeneratedRegex(@"\s+", RegexOptions.Compiled)]
    private static partial Regex WhitespacePattern();

    /// <summary>
    /// Normalizes a merchant name: lowercase, strip punctuation, collapse whitespace.
    /// Returns an empty string for null or whitespace-only input.
    /// </summary>
    public static string Normalize(string? name)
    {
        if (string.IsNullOrWhiteSpace(name))
            return string.Empty;

        var result = name.ToLowerInvariant();
        result = PunctuationPattern().Replace(result, " ");
        result = WhitespacePattern().Replace(result, " ");
        return result.Trim();
    }
}
